src/mame/taito/taito_f3.h

```diff
@@ -225,13 +225,20 @@ class taito_f3_state : public driver_device
 	u8 m_alpha_level_2bd = 0;
 	u8 m_alpha_level_3bs = 0;
 	u8 m_alpha_level_3bd = 0;
-	int m_alpha_level_last = 0;
+	u16 m_alpha_level_last = 0;
 	u16 m_width_mask = 0;
 	u8 m_twidth_mask = 0;
 	u8 m_twidth_mask_bit = 0;
 	std::unique_ptr<u8[]> m_tile_opaque_sp;
 	std::unique_ptr<u8[]> m_tile_opaque_pf[8];
-	u8 m_add_sat[256][256]{};
+	static constexpr std::array<std::array<u8, 256>, 256> m_add_sat = []
+	{
+		auto add_sat = std::array<std::array<u8, 256>, 256>{};
+		for (int i = 0; i < 256; i++)
+			for (int j = 0; j < 256; j++)
+				add_sat[i][j] = std::min(i + j, 255);
+		return add_sat;
+	}();
```

Comment:
You shouldn’t use the `m_` prefix for things that behave like constants or for class statics.  But that aside, is this table even worthwhile?  Often it’s more efficient to just calculate stuff like this on-the-fly with modern CPUs.

Reply:
You're right. I thought it was neat idea but it's probably best to generate it at runtime. Reverted.

Reply:
Have you tried removing `m_add_sat` altogether?  It’s only used in one function for three saturating additions:
```c++
inline void taito_f3_state::alpha_blend32_d(int alphas, u32 s)
{
	u8 *sc = (u8 *)&s;
	u8 *dc = (u8 *)&m_dval;
	dc[COLOR1] = m_add_sat[dc[COLOR1]][(alphas * sc[COLOR1]) >> 8];
	dc[COLOR2] = m_add_sat[dc[COLOR2]][(alphas * sc[COLOR2]) >> 8];
	dc[COLOR3] = m_add_sat[dc[COLOR3]][(alphas * sc[COLOR3]) >> 8];
}
```

You could change it to:
```c++
inline void taito_f3_state::alpha_blend32_d(int alphas, u32 s)
{
	u8 *sc = (u8 *)&s;
	u8 *dc = (u8 *)&m_dval;
	dc[COLOR1] = std::min<unsigned>(dc[COLOR1] + ((alphas * sc[COLOR1]) >> 8), 255U);
	dc[COLOR2] = std::min<unsigned>(dc[COLOR2] + ((alphas * sc[COLOR2]) >> 8), 255U);
	dc[COLOR3] = std::min<unsigned>(dc[COLOR3] + ((alphas * sc[COLOR3]) >> 8), 255U);
}
```

This will avoid keeping the 64KiB table hot.  It shouldn’t require conditional branches as the operations can all be represented as comparisons followed by conditional moves (or a vector compare/or/mask sequence if the compiler is smart enough to recognise it).  In many cases, modern CPUs perform better doing these kinds of calculations and conditional moves than fetching values from tables.

Reply:
Ah! Sorry, I misunderstood you. I removed m_add_sat and tried your change, seems good!